implement SO_BINDTODEVICE for inet protocols

RoadRunnr opened the pull request:
bind to device is needed to properly support VRF-Lite under Linux (see [VRF] for details).

This is a first draft to solicit comments. Documentation and tests are missing.

Is a change/feature like this acceptable?
Are the changes generally in the right place?

[VRF]: https://www.kernel.org/doc/Documentation/networking/vrf.txt

KennethL commented on lib/kernel/src/inet.erl:
```diff
@@ -296,6 +298,8 @@ getopts(Socket, Opts) ->
 	     [case OptionValue of
 		  {netns,Bin} ->
 		      {netns,binary2filename(Bin)};
+                  {bind_to_devive,Bin} ->
```
Error: Should be `bind_to_device` here

RaimoNiskanen commented:
I wonder why binding to an interface is needed?  I tried to read the VRF document, but did not get it.  On *BSD it seems this option does not exist and one should figure out the address to bind to to get bound to a particular interface...  Interface names are not portable between systems.  So why does not bind to an address cut it in this case?  It would be a Linux-only feature.

Furthermore I think testing for SO_BINDTODEVICE should be an AC_CHECK_DECL(s) configure test.  In inet_drv it is (almost) done so for SO_BSDCOMPAT but not for SO_PRIORITY, which I think is incorrect.

Other than that I think the changes are in the right places and seems correct.

RoadRunnr commented:
----- On Feb 17, 2017, at 3:20 PM, Raimo Niskanen <notifications@github.com> wrote: 

> I wonder why binding to an interface is needed? I tried to read the VRF
> document, but did not get it.

VRF-Lite is way to have multiple overlapping IP address ranges and overlapping routing domains. 
You can essential have multiple IP interfaces with the same IP address within the same network namespace. 

Obviously, when the IP address is no longer unique binding to it is no longer sufficient. Also, when you want to have your socket bound to a specific VRF instance, you have to use the BINDTODEVICE mechanism. 

> On *BSD it seems this option does not exist and one should figure out the
> address to bind to to get bound to a particular interface... Interface names
> are not portable between systems. So why does not bind to an address cut it in
> this case? It would be a Linux-only feature.

I think Linux is currently the only Erlang supported OS that implements a VRF mechanism. But the same applies to the 'netns' argument and that was added. 

> Furthermore I think testing for SO_BINDTODEVICE should be an AC_CHECK_DECL(s)
> configure test. In inet_drv it is (almost) done so for SO_BSDCOMPAT but not for
> SO_PRIORITY, which I think is incorrect.

> Other than that I think the changes are in the right places and seems correct.

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, [
> https://github.com/erlang/otp/pull/1326#issuecomment-280661836 | view it on
> GitHub ] , or [
> https://github.com/notifications/unsubscribe-auth/AAUkMaPKmGVYGXCbVFMd04yPR12Xun6yks5rdayfgaJpZM4L08nM
> | mute the thread ] .

RaimoNiskanen commented:
Well in that case this feature is acceptable. Please add documentation and tests, as you said, and also use a configure test for the declaration of SO_BINDTODEVICE.

RoadRunnr commented:
----- On Feb 20, 2017, at 10:02 AM, Raimo Niskanen <notifications@github.com> wrote: 

> Well in that case this feature is acceptable. Please add documentation and
> tests, as you said, and also use a configure test for the declaration of
> SO_BINDTODEVICE.

Could you explain why a configure test is needed? I can't really see the logic of defining a preprocessor symbol to signal the presence of another preprocessor symbol. This seems redundant.

RoadRunnr commented:
updated with tests and documentation.

I have added the configure test for SO_BINDTODEVICE as a second commit. I still can't see the need for it and would like to drop it.

RaimoNiskanen commented:
I did some digging since apparently for SO_BSDCOMPAT it did not always work with "#ifdef SO_BSDCOMPAT", and that must have been since it was not a preprocessor symbol, but an enumeration value.

A better way than to write your own configure code in this case should be to use AC_CHECK_DECLS([SO_BSDCOMPAT]).

**BUT** it seems that Posix defines (I will not swear to drink poison on that) that these animals should be *preprocessor symbols*, and on platforms that have them as enumeration values (Linux(?)) they should also be made available as preprocessor symbols with "#define SO_BINDTODEVICE SO_BINDTODEVICE".  Should this prove incorrect on some odd platform we can introduce a AC_CHECK_DECLS autoconf test then...

So it seems on some early Linux the SO_BSDCOMPAT flag was an enumeration value hence the cumbersome configure test for it, and that for e.g SO_BINDTODEVICE	 that should not be necessary.

Therefore you can remove the SO_BINDTODEVICE commit.

RoadRunnr commented:
updated, remove the configure test

RaimoNiskanen commented:
Now running in our daily tests.

A thought about the type of the interface name, though. If you compare to the option `netns`, maybe this option also should be a binary. That would be dodging from the character encoding problem since Unix/Linux essentially avoids talking about characters at the API level.

Not having the type as binary would bring forth the discussion about should you use `list_to_binary/1` or `unicode:characters_to_binary/1`?

RoadRunnr commented:
And -- surprise, surprise -- inet:getifaddrs/0 and the undocumented inet:getiflist/0 are broken for UTF-8 network device names....

With LANG set to UTF:

    $ env | grep UTF
    LANG=en_US.UTF-8

I can do this:

    # sudo ip link add öäü type dummy
    # sudo ip link set öäü up
    # sudo ip addr add 127.0.0.2/32 dev öäü
    # ip -br addr
    lo               UNKNOWN        127.0.0.1/8 ::1/128
    öäü           UNKNOWN        127.0.0.2/32 fe80::687b:1cff:feb4:335d/64


but:

    $ erl
    Eshell V8.2.2  (abort with ^G)
    1> inet:getiflist().
    {ok,["lo","Ã¶Ã¤Ã¼"]}
    2> inet:getifaddrs().
    {ok,[{"lo",
          [{flags,[up,loopback,running]},
           {hwaddr,[0,0,0,0,0,0]},
           {addr,{127,0,0,1}},
           {netmask,{255,0,0,0}},
           {addr,{0,0,0,0,0,0,0,1}},
           {netmask,{65535,65535,65535,65535,65535,65535,65535,
                     65535}}]},
         {"Ã¶Ã¤Ã¼",
          [{flags,[up,broadcast,running]},
           {hwaddr,[106,123,28,180,51,93]},
           {addr,{127,0,0,2}},
           {netmask,{255,255,255,255}},
           {broadaddr,{127,0,0,2}},
           {addr,{65152,0,0,0,26747,7423,65204,13149}},
           {netmask,{65535,65535,65535,65535,0,0,0,0}}]}]}

My guess is that no one is actually using UTF-8 characters in network device names.

RaimoNiskanen commented:
Oh dear... I took in `inet:getifaddrs/0`, before Unicode awareness increased above zero. :-(

You are probably right in that nobody uses non-7-bit-ASCII in network device names.  It is a greater problem in file names. But nevertheless, Murphy's law says that somebody will, and we should look forward to not introduce new bad behaviours.  Network interface names like other names (character arrays / strings) in Unix should be binaries (not containing zeros (and for file names not containing zeros nor slashes (except for paths))) since Unix does not give a hoot about character encoding.

So we also need a new variant of `inet:getifaddrs/0` (maybe `/1` with an option (list?)) that returns interface names as binaries.  But that would be another PR...

RaimoNiskanen commented:
The test cases do not work as they are now:
```
*** CT Error Notification 2017-03-01 12:24:40.416 ***
inet_SUITE:jog_bind_to_device_opt failed on line 1300
Reason: {badmatch,{error,einval}}

Full error description and stacktrace



=== Ended at 2017-03-01 12:24:40
=== Location: [{inet_SUITE,jog_bind_to_device_opt,1300},
              {inet_SUITE,simple_bind_to_device,1257},
              {test_server,ts_tc,1529},
              {test_server,run_test_case_eval1,1045},
              {test_server,run_test_case_eval,977}]
=== Reason: no match of right hand side value {error,einval}
  in function  inet_SUITE:jog_bind_to_device_opt/1 (inet_SUITE.erl, line 1300)
  in call from inet_SUITE:simple_bind_to_device/1 (inet_SUITE.erl, line 1257)
  in call from test_server:ts_tc/3 (test_server.erl, line 1529)
  in call from test_server:run_test_case_eval1/6 (test_server.erl, line 1045)
  in call from test_server:run_test_case_eval/9 (test_server.erl, line 977)
```
Is it so that extra privileges are required to set the option?  I recall reading that but can not find it.

The test case succeeds in reading the option but fails to set it to `<<>>`.

and:
```
*** CT Error Notification 2017-03-01 12:24:40.449 ***
gen_sctp:open failed on line 133
Reason: badarg

Full error description and stacktrace



=== Ended at 2017-03-01 12:24:40
=== Location: [{gen_sctp,open,133},
              {inet_SUITE,simple_bind_to_device_open,1289},
              {test_server,ts_tc,1529},
              {test_server,run_test_case_eval1,1045},
              {test_server,run_test_case_eval,977}]
=== Reason: bad argument
  in function  gen_sctp:open/1
     called as gen_sctp:open([{port,0},binary,{bind_to_device,<<"lo">>}])
  in call from inet_SUITE:simple_bind_to_device_open/1 (inet_SUITE.erl, line 1289)
  in call from test_server:ts_tc/3 (test_server.erl, line 1529)
  in call from test_server:run_test_case_eval1/6 (test_server.erl, line 1045)
  in call from test_server:run_test_case_eval/9 (test_server.erl, line 977)
```
Here it seems the SCTP code behaves differently (maybe a bug in itself) from the TCP and UDP code so it `exit(badarg)` instead of returning `{error,einval}`.

RoadRunnr commented:
> Is it so that extra privileges are required to set the option? I recall reading
> that but can not find it.

SO_BINDTODEVICE needs root or the CAP_NET_RAW capability (I did include that in the documentation update). 

> The test case succeeds in reading the option but fails to set it to <<>> .

In my tests this happens when the test process does not have sufficient rights. 
Setting the cap's like this: 

setcap cap_net_raw,cap_sys_admin,cap_sys_ptrace,cap_dac_read_search+epi beam.smp 

helps. 

> {test_server,run_test_case_eval1,1045}, {test_server,run_test_case_eval,977}]
> === Reason: bad argument in function gen_sctp:open/1 called as
> gen_sctp:open([{port,0},binary,{bind_to_device,<<"lo">>}]) in call from
> inet_SUITE:simple_bind_to_device_open/1 (inet_SUITE.erl, line 1289) in call
> from test_server:ts_tc/3 (test_server.erl, line 1529) in call from
> test_server:run_test_case_eval1/6 (test_server.erl, line 1045) in call from
> test_server:run_test_case_eval/9 (test_server.erl, line 977)

> Here it seems the SCTP code behaves differently (maybe a bug in itself) from the
> TCP and UDP code so it exit(badarg) instead of returning {error,einval} .

Yet, there is code in gen_sctp that converts the `{error, einval}` into a `exit(badarg)`

RaimoNiskanen commented:
Allright, it was in the documentation I had read it. Did not think of that ;-(

Well, the problem is that our daily build system builds and installs Erlang/OTP on about 60 test machines of which close to 60% are Linux ranging from kernel version 2.6 to 4.2.

The test machines are controlled by SSH login as an unprivileged user and hence can not use root privileges other than possibly through sudo but then we have to reconfigure all Linux machines and rewrite the daily build scripts to use sudo to elevate capabilities on some machines.

So we need a framework to handle how to and on which machines to elevate privileges and have to enable that on a subset of the machines.

This is on our todo list, but we can not even after that has been done have a test case that fails out of the box on an installation as a regular user that can not give elevated capabilities to the installed Erlang VM.

The SCTP problem is also something we should fix, but to 20.0 since it is a behavioural change, and we can not have a test case fail on all platforms until then. There are enough failing test cases already.

So I suggest that you fix the `jog_bind_to_device_opt` to be skipped on platforms without elevated privileges.

And comment out the `gen_sctp` check in `simple_bind_to_device_open`, with fat `XXX` comments and reminders.  The explicit conversion from `{error,einval}` into `exit(badarg)` was a mistake (of me, I think). It is now obvious that it should behave like `gen_tcp` and `gen_udp`...

IngelaAndin commented:
ping

RoadRunnr commented:
----- On Apr 20, 2017, at 2:40 PM, Ingela Andin <notifications@github.com> wrote: 

> ping

The last comment got somehow lost in my mail filter. 

I have reworked the test suite to first check if bind_to_device should be supported 
on this OS version and the to check that we have sufficient permissions. That 
way we don't have to rely on the error from the open call to check if it should work or not. 
Tested as root user and as unprivileged user. 

I think that should fix both the permission and sctp problem.

RaimoNiskanen commented:
I just read the code... Deep stuff, man!  Glad I did not have to figure that out.

RaimoNiskanen commented:
No regression test failures in our daily builds.  Merged to 'master'.  Thank you for your contribution!
